File: frontend/src/api/disruptions.ts
import { apiFetch, ApiError, readError } from './client'

export type DisruptionImpactClass = {
  assignment_id: number
  meeting_index: number
  course_code: string | null
  course_title: string | null
  cohort_code: string | null
  cohort_size: number
  weekday: string
  start_period: string
  end_period: string
  room_code: string | null
  lecturer_name: string | null
}

export type DisruptionImpact = {
  classes_affected: number
  students_affected: number
  published: boolean
  classes: DisruptionImpactClass[]
}

export type DisruptionRecord = {
  id: number
  code: string
  session_id: number
  kind: string
  room_id: number | null
  lecturer_id: number | null
  resource_label: string
  reason: string
  description: string | null
  severity: string
  starts_on: string
  ends_on: string
  start_period: string | null
  end_period: string | null
  status: string
  reported_by: number
  reporter_name: string | null
  reporter_role: string | null
  block_id: number | null
  classes_affected: number
  students_affected: number
  created_at: string
  updated_at: string
  impact: DisruptionImpact | null
}

export type DisruptionSummary = {
  active: number
  scheduled: number
  open: number
  in_review: number
  room_active: number
  lecturer_active: number
  classes_affected: number
  students_affected: number
  published: boolean
}

export type DisruptionWrite = {
  kind: string
  room_id?: number | null
  lecturer_id?: number | null
  reason: string
  description?: string | null
  severity?: string
  starts_on: string
  ends_on: string
  start_period?: string | null
  end_period?: string | null
  create_block?: boolean
}

async function requestJson<T>(path: string, init?: RequestInit): Promise<T> {
  const response = await apiFetch(path, init)
  if (!response.ok) {
    throw new ApiError(response.status, await readError(response))
  }
  if (response.status === 204) {
    return undefined as T
  }
  return (await response.json()) as T
}

export function fetchDisruptions(params: Record<string, string | undefined> = {}) {
  const search = new URLSearchParams()
  for (const [key, value] of Object.entries(params)) {
    if (value) {
      search.set(key, value)
    }
  }
  const query = search.toString()
  return requestJson<DisruptionRecord[]>(`/api/disruptions${query ? `?${query}` : ''}`)
}

export function fetchDisruptionSummary() {
  return requestJson<DisruptionSummary>('/api/disruptions/summary')
}

export function fetchDisruption(id: number) {
  return requestJson<DisruptionRecord>(`/api/disruptions/${id}`)
}

export function fetchDisruptionImpact(id: number) {
  return requestJson<DisruptionImpact>(`/api/disruptions/${id}/impact`)
}

export function previewDisruption(body: DisruptionWrite) {
  return requestJson<DisruptionImpact>('/api/disruptions/preview', {
    method: 'POST',
    body: JSON.stringify(body),
  })
}

export function createDisruption(body: DisruptionWrite) {
  return requestJson<DisruptionRecord>('/api/disruptions', {
    method: 'POST',
    body: JSON.stringify(body),
  })
}

export function patchDisruption(id: number, body: Record<string, unknown>) {
  return requestJson<DisruptionRecord>(`/api/disruptions/${id}`, {
    method: 'PATCH',
    body: JSON.stringify(body),
  })
}

File: backend/app/schemas/disruptions.py
from datetime import date

from pydantic import BaseModel, ConfigDict, Field


class DisruptionImpactClass(BaseModel):
    assignment_id: int
    meeting_index: int
    course_code: str | None = None
    course_title: str | None = None
    cohort_code: str | None = None
    cohort_size: int = 0
    weekday: str
    start_period: str
    end_period: str
    room_code: str | None = None
    lecturer_name: str | None = None


class DisruptionImpact(BaseModel):
    classes_affected: int
    students_affected: int
    published: bool
    classes: list[DisruptionImpactClass] = Field(default_factory=list)


class DisruptionIn(BaseModel):
    kind: str
    room_id: int | None = None
    lecturer_id: int | None = None
    reason: str = Field(min_length=1, max_length=255)
    description: str | None = Field(default=None, max_length=2000)
    severity: str = "medium"
    starts_on: date
    ends_on: date
    start_period: str | None = None
    end_period: str | None = None
    create_block: bool = False


class DisruptionPatch(BaseModel):
    status: str | None = None
    reason: str | None = Field(default=None, min_length=1, max_length=255)
    description: str | None = Field(default=None, max_length=2000)
    severity: str | None = None
    starts_on: date | None = None
    ends_on: date | None = None
    start_period: str | None = None
    end_period: str | None = None


class DisruptionOut(BaseModel):
    id: int
    code: str
    session_id: int
    kind: str
    room_id: int | None
    lecturer_id: int | None
    resource_label: str
    reason: str
    description: str | None
    severity: str
    starts_on: date
    ends_on: date
    start_period: str | None
    end_period: str | None
    status: str
    reported_by: int
    reporter_name: str | None
    reporter_role: str | None
    block_id: int | None
    classes_affected: int
    students_affected: int
    created_at: str
    updated_at: str
    impact: DisruptionImpact | None = None

    model_config = ConfigDict(from_attributes=True)


class DisruptionSummaryOut(BaseModel):
    active: int
    scheduled: int
    open: int
    in_review: int
    room_active: int
    lecturer_active: int
    classes_affected: int
    students_affected: int
    published: bool
